### ISSUE TITLE
Feature: Increased resilience on `Fetch()` and `Parse()` in case of invalid key in JWKS-endpoint?

### ISSUE THREAD

**Issue by mzc-lunar** [lestrrat-go/jwx#567]
**Intro**
Hi

First of all, thank you for the work you've put into the jwx-project. It is clear that a lot of effort has gone into getting all the details of all the JWX packages to behave in a nice way that is easy to understand and such that it follows the various protocols to the letter.

However, in using your libraries we ended up in an interesting pickle this past week, which we thought we would bring by you to get your thoughts on it.

Of course we understand that we obviously can't _demand_ any work on this particular issue but we thought we would bring it up to see if it happened to be something that could fit into a future version of the JWK-package.

**Abstract**
The specific issue arrived when the OIDC identity-provider we integrate up against added an invalid key to their JWKS-endpoint. This happened to be an RSA-key with `"alg":"enc"` used for encrypted requests.
This feature is not something we are planning to use, but the issue came since the specific public key was missing the `"e"` and `"n"` properties, required by this [RFC](https://datatracker.ietf.org/doc/html/rfc7518#section-6.3.1), the key failed to unmarshal, causing our key-validation to fail.

The root of the issue is of course not with the JWK-library provided by you, but from our identity-provider _not following the protocol_. However, we are wondering if it could be an idea to make the `AutoRefresh.Fetch()` function a little more resilient towards this kind of issue?

I've put a few of my initial ideas in the next section, to illustrate how I imagine this could be done.

I haven't written direct failing unit-tests to illustrate my meaning, but I have the following [gist](https://gist.github.com/mzc-lunar/ff7d0f26741f2dcb0c15e654a1578d3d) which shows a super small playground-example of how the parsing
fails when trying to parse the key with kid `A2E10A6BAF4E43E86273F57F218A44B824203176` as it is missing param `"e"` and `"n"`. So from an API-point of view the consumer views it as the `.Fetch()` function failing in my specific instance, but one could also argue that I am requesting a change to the `.Parse()` which is also exposed to consumers.

**Describe the proposed solution/change**
Possible ideas could be to have a version of fetch which only returns keys with certain KIDs, or have it "optimistically" return keys which are a valid `Set` type?

I don't think either of these should be the default behaviour, but the issue is that the JWKS-endpoint is of course something external to ones system, so a possibility of being slightly defensive about its contents could be valuable.

**Analysis**
From looking at the code-base, I think either solution is doable by making new versions of the `.Fetch()` function which behaves slightly differently in the layers where the endpoint is called and the result is unmarshalled.
Personally, I think I would lean towards the approach of giving the KIDs as a slice in to a `.FetchKIDs()` function or something like that.  If the KIDs are not available you could then return a specific error and the consumer could decide whether to call `.Fetch()` afterwards in case the keys have been rolled.

To be completely fair, it is of course also a possibility to manually parse the interesting keys using the `.Parse()` functionality which fixes it in our end, without requiring any changes in the JWK-library. This would make it possible for us to only focus on the keys we are interested in, while ignoring potentially flawed keys. I just really enjoy the idea of "not having to think about the JWKS-endpoint at all", which is why we used the `.Fetch()` function in the first place.

**Comment by mzc-lunar** [comment]
Hi @lestrrat Thank you for the quick response.

First off, your four assumptions on my initial post are correct and I think you understand our issue perfectly fine.
Let me elaborate a little more about my reasoning behind why I think the suggested feature would be quite powerful.
I think there is something really useful about a consumer being able to call something like

```GO
autoRefresher.Fetch(ctx, jwksURL)
```

on an `AutoRefresher` object and not having to "think about what happens behind the scenes" regarding whether the keys are retrieved from cache or from the JWKS-endpoint.

However, this is of course vulnerable to potential issues with the JWKS-endpoint and I'm just wondering if it wouldn't be possible to build in some resilience without making too many compromises.

to better come up with a specific suggestion to what the new API would look like, I've had a look at what our token-validation code looks like specifically. My guess is that a lot of uses will look something like the above call to `autoRefresher.Fetch()` closely followed by a call to `jwt.Parse()` which takes an access token as input (at least we do something a long those lines). So maybe it's not necessary to have the new input be a slice of `kid`s but rather a single kid as we are validating a specific access token when we want to get a key(set). I haven't looked into exactly how to get the header from a byte-array representation of an access token using the jwt-package, but I imagine it's possible using your library?

so forgive my lack of knowledge on what this would look like using the lib, but using pseudo-code I'd do something like

```GO

kid, err := jwt.GetKid(accessToken) // in reality this is done with some other way
if err != nil {
   // handle err
}

keys, err := autoRefresher.FetchKid(ctx, jwksURL, kid) // this is the new functionality I'd hope for
if err != nil {
   // handle err
}

jwt.Parse(accessToken,
    jwt.WithKeySet(keys),
    ...
)
```


In this scenario, the `FetchKid()` would only parse the JWK corresponding to the given `kid` twice, meaning for a JWKS-endpoint with with `n` keys, it would parse a worst-case of `n+1` json-objects, and possibly even less since I see no reason to keep parsing json after you find the key you are looking for.

Furthermore, I'd say the behaviour of this function would be that if we can parse the single key we are looking for as a valid JWK we return that, but if it is missing or invalidly formatted a well-defined error is returned.

My idea here is that this case would anwser both the performance issue and the behaviour issue you have pointed out.

I've of course here assumed that we have a `jwt.GetKid()` function available in the library, which we might not have out of the box - but I imagine that wouldn't be the hard part of the new feature I am suggestion.


Thoughts?

**Comment by lestrrat** [comment]
@mzc-lunar So if I read this correctly, your proposed feature is more of a `ExtractSingleKeyFromSet()` (I know it sounds bad, just trying to get my point across), to extract a single key out of a set, and parse that key only.

I haven't thought this through yet, but I _think_ I want to separate the extraction of a single key (JSON) and parsing said JSON in two steps. Bundling everything in a function sounds like a sure path to disaster later on in the future.

**Comment by mzc-lunar** [comment]
Yes, I think that is accurate. The primary reason for not doing something our-selves is the fact that using the jwk-lib gives us fetching with caching combined with parsing of the JWK, which I think is quite nice. Your argument with seperating the functionality would then be to have a function `FetchKey()` and then following that with a call to `Parse()` ?

**Comment by lestrrat** [comment]
Again, this is not a rejection of your proposal, but as far as utilizing the fetch+caching goes, you could still do this:

1. periodically fetch http://example.com/jwks.json (the original source)
2. use jq to do the filtering, save it somewhere, store it where you can create a private http server (it could even by S3 or something similar) as http://internal-location/jwks.json
3. point JWK to http://intranal-location/jwks.json

Just a thought.

Meanwhile, still thinking about it.

**Comment by mzc-lunar** [comment]
Yes I think this is a good point. We are looking into doing something similar to that. The issue is that the more disjoint we become from the "actual" JWKS-endpoint the more we risk not noticing if a key is rolled (possibly due to the private key being compromised). So by keeping an extra cache of the JWKS-endpoint result we are trading resilience for a loss of security in edge-cases like that.

I realise I'm pointing out edge cases here. But before this, I wouldn't have thought an addition of a key we don't care about to the JWKS-endpoint could cause our whole OIDC-flow to fail 🙈 

I appreciate your thoughts on the topic, and I won't blame you if you end up deciding that this is not worth the effort - but I thought I would give it a go nonetheless 😄

**Comment by lestrrat** [comment]
@mzc-lunar so here's one way I thought of https://github.com/lestrrat-go/jwx/pull/570/files#diff-40f65963e86f0e5ec7c15dd282ccb94f2feff9934e698bf2f51c98598ec5b45bR1965

Let me know what you think. Meanwhile it's bed time where I live. I will check back tomorrow.

**Comment by lestrrat** [comment]
@mzc-lunar Here's another possible solution I thought of: Just allow jwk.Parse to ignore errors: https://github.com/lestrrat-go/jwx/pull/574/files#diff-40f65963e86f0e5ec7c15dd282ccb94f2feff9934e698bf2f51c98598ec5b45bR1966

**Comment by mzc-lunar** [comment]
Hi again @lestrrat that was some fast coding 🎉 

I think for my team we would personally prefer option two as it would probably be the the most "hands off" approach from our end which probably means smaller risk of errors. I kind of like that the idea of being able to set `jwk.WithIgnoreParseError(true)` and then we would just get "any valid key" from JWKS back 👏

**Comment by lestrrat** [comment]
Okay, in that case let's go with the "Allow users to shoot themselves in the foot" approach. I still need to write documentation and test a bit more before releasing.

**Comment by lestrrat** [comment]
@mzc-lunar #574 is ready, can you please take it for a whirl and see if it does the job for you. If it works for you, we'll merge to main. Then we will let it sit for a few days in main to give ourselves one last chance to fix anything if necessary before releasing.

**Comment by mzc-lunar** [comment]
@lestrrat I'll have a look 👍  Thank you again for the quick turnaround on this

**Comment by mzc-lunar** [comment]
@lestrrat We've been running it in our development environment and it seems to be running fine. The option it self hasn't been necessary at the moment as the JWKS-endpoint we use is currently valid. But from looking at the PR I believe it should work fine if the endpoint is messed up again. So all in all, we're very happy with the result.

Coincidentally, I learned that our organization is currently looking into possibly supporting "important open-source projects" that we rely on. I have no say in who possibly gets any support and who doesn't, but I have nominated the JWX-project as something worth supporting as I a lot of care being put into the project. I can't make any promises of course, but I thought I would tell you nonetheless that I hope some support is coming your way from us.
